hkmc2/shared/src/main/scala/hkmc2/codegen/Lifter.scala

```diff
@@ -0,0 +1,1504 @@
+            N, Diagnostic.Source.Compilation
+          ))
+          ignored += l
+          v
+        case _ => super.applyValue(v)
+    walker2.applyDefn(d)
+
+    (ignored, modules, objects)
+
+  extension (b: Block)
+    private def floatOut(ctx: LifterCtx) =
+      b.floatOutDefns(preserve = defn => ctx.isModOrObj(defn.sym) || ctx.ignored(defn.sym))
+      
+
+  def createLiftInfoCont(d: Defn, parentCls: Opt[ClsLikeDefn], ctx: LifterCtx): Map[BlockMemberSymbol, LiftedInfo] =
+    val AccessInfo(accessed, _, refdDefns) = ctx.getAccesses(d.sym)
+
+    val inScopeRefs = refdDefns.intersect(ctx.inScopeDefns(d.sym))
+
+    val includedCaptures = ctx.prevFnLocals.reqCapture
+      .intersect(accessed)
+      .map(sym => ctx.lookup(sym).get)
+      .toList.sortBy(_.uid)
+
+    val refMod = inScopeRefs.intersect(ctx.modLocals.keySet)
+    val includedLocals = ((accessed -- ctx.prevFnLocals.reqCapture) ++ refMod).toList.sortBy(_.uid)
+    val clsCaptures: List[InnerSymbol] = ctx.prevClsDefns.map(_.isym)
+    val refBms = inScopeRefs.intersect(ctx.ignoredDefns).toList.sortBy(_.uid)
+
+    val modLocal = d match
+      case c: ClsLikeDefn if modOrObj(c) => parentCls match
+        case None => S(VarSymbol(Tree.Ident(c.sym.nme + "$")))
+        case Some(value) => S(TermSymbol(syntax.ImmutVal, S(value.isym), Tree.Ident(c.sym.nme + "$")))
+      case _ => N
+
+    if ctx.ignored(d.sym) ||
+      (includedCaptures.isEmpty && includedLocals.isEmpty && clsCaptures.isEmpty && refBms.isEmpty) then
+      d match
+        case f: FunDefn =>
+          createLiftInfoFn(f, ctx)
+        case c: ClsLikeDefn =>
+          createLiftInfoCls(c, ctx)
+        case _ => Map.empty
+    else
+      val fakeCtorBms = d match
+        case c: ClsLikeDefn if !modLocal.isDefined => S(BlockMemberSymbol(d.sym.nme + "$ctor", Nil))
+        case _ => N
+
+      val singleCallBms = BlockMemberSymbol(d.sym.nme + "$", Nil)
+
+      val info = LiftedInfo(
+        includedCaptures, includedLocals, clsCaptures,
+        refBms, fakeCtorBms, singleCallBms
+      )
+      
+      d match
+        case f: FunDefn =>
+          createLiftInfoFn(f, ctx) + (d.sym -> info)
+        case c: ClsLikeDefn =>
+          createLiftInfoCls(c, ctx) + (d.sym -> info)
+        case _ => Map.empty
+  
+  def createLiftInfoFn(f: FunDefn, ctx: LifterCtx): Map[BlockMemberSymbol, LiftedInfo] =
+    val defns = ctx.nestedDefns(f.sym)
+    defns.flatMap(createLiftInfoCont(_, N, ctx.addFnLocals(ctx.usedLocals(f.sym)))).toMap
+
+  def createLiftInfoCls(c: ClsLikeDefn, ctx: LifterCtx): Map[BlockMemberSymbol, LiftedInfo] =
+    val defns = c.preCtor.floatOut(ctx)._2 ++ c.ctor.floatOut(ctx)._2
+    val newCtx = if c.k is syntax.Mod then ctx else ctx.addClsDefn(c)
+    defns.flatMap(f => createLiftInfoCont(f, S(c), newCtx)).toMap
+      ++ c.methods.flatMap(f => createLiftInfoFn(f, newCtx))
+
+  def refRewriter(ctx: LifterCtx, ctorCls: Opt[ClsLikeDefn]) = 
```

Comment:
Same remark as for `rewriteBlk`. Just food for thought – maybe some of these transformations could be defined more simply instead of being nested in definitions.